internal/impl/mssqlserver/replication/stream.go

```diff
@@ -0,0 +1,411 @@
+	return ct.rows.Close()
+}
+
+// mapValsToChange maps the values from vals to the dst out parameter.
+func (ct *changeTableRowIter) mapValsToChange(vals []any, dst *change) error {
+	for i, c := range ct.cols {
+		v := *(vals[i].(*any))
+		switch c {
+		case "__$start_lsn":
+			if b, ok := v.([]byte); ok {
+				dst.startLSN = b
+			} else {
+				return errors.New("failed to map 'start_lsn' column from change table")
+			}
+		case "__$end_lsn":
+			// "In SQL Server 2012 (11.x), this column is always NULL."
+			// https://learn.microsoft.com/en-us/sql/relational-databases/system-tables/cdc-capture-instance-ct-transact-sql?view=sql-server-ver16
+			if b, ok := v.([]byte); ok {
+				dst.endLSN = b
+			} else if v == nil {
+				dst.endLSN = nil
+			} else {
+				ct.log.Warnf("failed to map 'end_lsn' column from change table")
+			}
+		case "__$update_mask":
+			if b, ok := v.([]byte); ok {
+				dst.updateMask = b
+			} else {
+				return errors.New("failed to map 'update_mask' column from change table")
+			}
+		case "__$operation":
+			switch x := v.(type) {
+			case int64:
+				dst.operation = int(x)
+			case int32:
+				dst.operation = int(x)
+			default:
+				return errors.New("failed to map 'operation' column from change table")
+			}
+		case "__$command_id":
+			switch x := v.(type) {
+			case int64:
+				dst.commandID = int(x)
+			case int32:
+				dst.commandID = int(x)
+			default:
+				return errors.New("failed to map 'command_id' column from change table")
+			}
+		case "__$seqval":
+			if b, ok := v.([]byte); ok {
+				dst.seqVal = b
+			} else {
+				return errors.New("failed to map 'seqval' column from change table")
+			}
+		default:
+			dst.columns[c] = v
+		}
+	}
+	return nil
+}
+
+// ChangePublisher is responsible for handling and processing of a replication.MessageEvent.
+type ChangePublisher interface {
+	Publish(ctx context.Context, msg MessageEvent) error
+}
+
+// ChangeTableStream tracks and streams all change events from the configured change tables tracked in tables.
+type ChangeTableStream struct {
+	tables    []UserTable
+	publisher ChangePublisher
+	log       *service.Logger
+}
+
+// NewChangeTableStream creates a new instance of NewChangeTableStream, responsible for paging through change events
+// based on the tables param.
+func NewChangeTableStream(tables []UserTable, publisher ChangePublisher, logger *service.Logger) *ChangeTableStream {
+	s := &ChangeTableStream{
+		tables:    tables,
+		publisher: publisher,
+		log:       logger,
+	}
+	return s
+}
+
+// ReadChangeTables streams the change events from the configured SQL Server change tables.
+func (r *ChangeTableStream) ReadChangeTables(ctx context.Context, db *sql.DB, startPos LSN) error {
+	r.log.Infof("Starting streaming %d change table(s)", len(r.tables))
+	var (
+		fromLSN LSN // load last checkpoint; nil means start from beginning in tables
+		toLSN   LSN // often set to fn_cdc_get_max_lsn(); nil means no upper bound
```

Comment:
We have some naming discrepancy, could we use startLSN and endLSN here too?

Reply:
Updated in f0fb4b859